kep/61-Trimaran-real-load-aware-scheduling/README.md

```diff
@@ -0,0 +1,588 @@
+- Safe Balancing Plugin: It is a node sorting plugin that sorts nodes base on both the mean and the standard deviation of node resource utilization. It aims to balance not only the average load but also the risk caused by load variations.
+
+
+### BestFitBinPack Plugin
+
+
+#### Score plugin
+
+This plugin would extend the Score extension point. K8s scheduler framework calls the Score function for each node separately when scheduling a pod.
+
+Following is the algorithm:
+
+**Algorithm**
+
+1. Get the utilization of the current node to be scored. Call it A.
+2. Calculate the current pod's total CPU requests and overhead. Call it B.
+3. Calculate the expected utilization if the pod is scheduled under this node by adding i.e. U = A + B.
+4. If U &lt;= X%, return (100 - X)U/X + X as the score
+5. If X% &lt; U &lt;= 100%, return 50(100 - U)/(100 - X)
+6. If U > 100%, return 0
+
+For example, let’s say we have three nodes X, Y, and Z, with four cores each and utilization 1, 2, and 3 cores respectively. For simplicity, let’s assume our pod to be scheduled has 0 cores CPU requests and overhead. Let X = 50%.
+
+Utilization of each node:
+
+
+```
+Ux → (1/4)*100 = 25
+Uy → (2/4)*100 = 50
+Uz → (3/4)*100 = 75
+```
+
+
+The score of each node :
+
+
+```
+X → (100 - 50)*25/50 + 50 =  75
+Y → (100 - 50)*50/50 + 50 = 100
+Z → 50 * (100 - 75)/(100 - 50) = 25
+```
+
+
+In the algorithm above, 50% is the target utilization rate we want to achieve on all nodes. We can be less aggressive by reducing it to 40% so that it has much lesser chances of going over 50% during spikes or unexpected loads. 
+So in general to achieve X% utilisation, X - 10 value is recommended in practice.
+
+In the 2nd step of the algorithm, one variant uses the current pod's total CPU limits instead of requests, to have a stronger upper bound of expected utilization.
+
+The X% threshold value for utilisation will be made configurable via plugin argument.
+
+**Algorithm Analysis**
+
+<img src="images/bfbp-graph.png" alt="bfbp-graph" width="600" height="400"/>
+
+
+The above is a plot of the piecewise function outlined in the algorithm. The key observations here are manifold:
+
+
+
+1. As the utilization goes from 0 to 50%, we pack pods on those nodes by favoring them.
+2. The nodes are penalized linearly as the utilization goes beyond 50%, to spread the pod amongst those "hot" nodes.
+3. The positive slope begins from 50 and not from 0 because the range of score is from 0-100, and we want to maximize our score output for nodes we wish to favor so that the score is more substantial and other plugins do not affect it much.
+4. There is a break in the graph with a high drop due to the "penalty" we have.
+
+**Plugin Config Args**
+```go
+type PluginArgs struct {
+    TargetCPUUtilisation    int
+    DefaultCPURequests      v1.ResourceList
+}
+```
+
+
+### Safe Balancing Plugin
```

Comment:
What about "Safe Utilization"?

Reply:
@alculquicondor, sounds good too. We refer to balancing as it balances the risks and the risk denotes the fluctuating utilization. How about `Risk Balancing plugin`?

Reply:
But what is it balancing? Honestly, it's going to be hard to come up with a good short name. But at least it should have a similar suffix to the other plugin, given that it uses the same source of information and it's trying to do something about the Load or Utilization.

Reply:
I like “risk balancing”. The suggestion “safe utilization” may imply that we’re determining a value of utilization that is safe, and that’s not true.

Reply:
"Safe Loading"?

Reply:
Will rename the `Safe Balancing` plugin to `LoadVariationRiskBalancing` plugin.